numpy/core/src/multiarray/abstractdtypes.c

```diff
@@ -174,7 +174,7 @@ int_common_dtype(PyArray_DTypeMeta *NPY_UNUSED(cls), PyArray_DTypeMeta *other)
 static PyArray_DTypeMeta *
 float_common_dtype(PyArray_DTypeMeta *cls, PyArray_DTypeMeta *other)
 {
-    if (other->legacy && other->type_num < NPY_NTYPES) {
+    if (NPY_DT_is_legacy(other) && other->type_num < NPY_NTYPES) {
```

Comment:
Would work too.  I am still trying to not assign type numbers at all to new-style dtypes.  But even then, maybe I should assign them something like `INT_MAX` rather than an invalid one like `-1` or `0`.